cluster-api/pkg/controller/machine/queue.go

```diff
@@ -0,0 +1,89 @@
+package machine
+
+import (
+	"time"
+	"github.com/golang/glog"
+	"k8s.io/apimachinery/pkg/util/runtime"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"github.com/kubernetes-incubator/apiserver-builder/pkg/controller"
+)
+
+func (c *MachineController) RunAsync(stopCh <-chan struct{}) {
+	for _, q := range c.Informers.WorkerQueues {
+		c.StartWorkerQueue(q,stopCh)
+	}
+	controller.GetDefaults(c.controller).Run(stopCh)
+}
+
+// StartWorkerQueue schedules a routine to continuously process Queue messages
+// until shutdown is closed
+func (c *MachineController) StartWorkerQueue(q *controller.QueueWorker, shutdown <-chan struct{}) {
+	defer runtime.HandleCrash()
+	glog.Infof("Start %s Queue", q.Name)
+
+	// Every second, process all messages in the Queue until it is time to shutdown
+	go wait.Until(func(){c.ProcessAllMessages(q)}, time.Second, shutdown)
+
+	go func() {
+		<-shutdown
+
+		// Stop accepting messages into the Queue
+		glog.V(1).Infof("Shutting down %s Queue\n", q.Name)
+		q.Queue.ShutDown()
+	}()
+}
+
+// ProcessAllMessages tries to process all messages in the Queue
+func (c *MachineController) ProcessAllMessages(q *controller.QueueWorker) {
+	for done := false; !done; {
+		// Process all messages in the Queue
+		done = c.ProcessMessage(q)
+	}
+}
+
+// ProcessMessage tries to process the next message in the Queue, and requeues on an error
+func (c *MachineController) ProcessMessage(q *controller.QueueWorker) bool {
+	key, quit := q.Queue.Get()
+	if quit {
+		// Queue is empty
+		return true
+	}
+
+	// Create sync channel if needed.
+	syncChan, ok := c.controller.syncChans[key.(string)]
```

Comment:
`q.Queue.Get()` already provides the needed synchronizations to operate on a single item.
Having those extra channels complicate the logic and make it harder to test. [sample-controller](https://github.com/kubernetes/sample-controller/blob/32ad1389a8f5758383883cc8413ea3b394fcdf78/controller.go#L152-L241)  can be used as a reference implementation for this kind of behavior.

Reply:
Thanks for pointing this out!